services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusService.kt

```diff
@@ -32,6 +58,264 @@ class HakemusService(
+     * the yhteystieto is not present in the application, it means that the request adds a new
+     * yhteystieto, and it cannot have an id. If the yhteystieto is not present in the request, it
+     * means that the application yhteystieto is removed.
+     */
+    private fun assertYhteystietoValidity(
+        applicationId: Long,
+        rooli: ApplicationContactType,
+        hakemusyhteystietoEntity: HakemusyhteystietoEntity?,
+        customerWithContacts: CustomerWithContactsRequest?
+    ) {
+        if (
+            hakemusyhteystietoEntity != null &&
+                customerWithContacts != null &&
+                customerWithContacts.customer.yhteystietoId != hakemusyhteystietoEntity.id
+        ) {
+            // ids don't match
+            throw InvalidHakemusyhteystietoException(
+                applicationId,
+                rooli,
+                hakemusyhteystietoEntity.id,
+                customerWithContacts.customer.yhteystietoId
+            )
+        }
+        if (
+            hakemusyhteystietoEntity == null &&
+                customerWithContacts?.customer?.yhteystietoId != null
+        ) {
+            // new yhteystieto has an id when it shouldn't
+            throw InvalidHakemusyhteystietoException(
+                applicationId,
+                rooli,
+                null,
+                customerWithContacts.customer.yhteystietoId
+            )
+        }
+    }
+
+    /** Assert that the contacts are users of the hanke. */
+    private fun assertYhteyshenkilotValidity(
+        hanke: HankeEntity,
+        newHankekayttajaIds: Set<UUID>,
+        customMessageOnFailure: (Set<UUID>) -> String
+    ) {
+        val currentHankekayttajaIds =
+            hankeKayttajaService.getKayttajatByHankeId(hanke.id).map { it.id }.toSet()
+        val newInvalidHankekayttajaIds = newHankekayttajaIds.minus(currentHankekayttajaIds)
+        if (newInvalidHankekayttajaIds.isNotEmpty()) {
+            throw InvalidHakemusyhteyshenkiloException(
+                customMessageOnFailure(newInvalidHankekayttajaIds)
+            )
+        }
+    }
+
+    /** Assert that the geometries are compatible with the hanke area geometries. */
+    private fun assertGeometryCompatibility(
+        hankeId: Int,
+        areas: List<ApplicationArea>,
+        customMessageOnFailure: (ApplicationArea) -> String
+    ) {
+        areas.forEach { area ->
+            if (!geometriatDao.isInsideHankeAlueet(hankeId, area.geometry))
+                throw ApplicationGeometryNotInsideHankeException(customMessageOnFailure(area))
+        }
+    }
+
+    /** Update the hanke areas based on the update request work areas. */
+    private fun updateHankealueet(hankeEntity: HankeEntity, updateRequest: HakemusUpdateRequest) {
+        val hankealueet =
+            updateRequest.areas?.mapIndexed { i, area ->
+                area.toNewHankealue(i, updateRequest.startTime!!, updateRequest.endTime!!)
+            } ?: emptyList()
```

Comment:
This, including `toNewHankealue`, is very close to `HankealueService.createHankealueetFromCableReport`. That could be refactored to take `areas: List<ApplicationArea>?` as its parameter and used in both cases.

Reply:
Nicely spotted, thanks!

Reply:
The version taking a CableReportApplicationData could be left as an overloaded method since it's called from a couple of places already.
```
        fun createHankealueetFromCableReport(
            cableReportData: CableReportApplicationData
        ): List<NewHankealue> =
            createHankealueetFromApplicationAreas(
                cableReportData.areas,
                cableReportData.startTime,
                cableReportData.endTime
            )
```

Reply:
Did that.